### PR TITLE
Initiate payment for approved orders (TTVA-82)

### PR DESCRIPTION
**Pull request by jopesy**
- Allow changing reservation state from `REQUESTED` to `WAITING_FOR_PAYMENT` via the API
- Initiate the Ceepos payment process when the state is changed from `REQUESTED` to `WAITING_FOR_PAYMENT` via the API, i.e. when a paid reservation is approved. The return URL is required to be present in the `PUT` request payload.
- Send email to the customer when a paid reservation is approved. The payment link (`Order.payment_link`) can be included in the email template as a template variable.

Refs TTVA-82, TTVA-86

### CODE REVIEW

File: resources/api/reservation.py [hunk 1]
```diff
@@ -768,8 +768,27 @@ def perform_create(self, serializer):
...
+        if (
+            order
+            and old_state == Reservation.REQUESTED
+            and new_state == Reservation.WAITING_FOR_PAYMENT
+        ):
+            if not payment_return_url:
+                raise ValidationError(_("Return URL is required to initiate the payment"))
+            else:
+                # Initiate the payment process
+                from payments.providers import get_payment_provider
```
Comment (danjacob-anders) [inline resources/api/reservation.py]: Does importing `get_payment_provider` cause a circular import if imported at top of module? If not, then import can be moved out of the method.

File: payments/tests/conftest.py [hunk 2]
```diff
@@ -37,6 +37,35 @@ def two_hour_reservation(resource_in_unit, user):
     )
 
 
+@pytest.fixture
+def requested_reservation_with_order(resource_in_unit, user):
+    reservation = Reservation.objects.create(
+        resource=resource_in_unit,
+        begin=datetime.datetime(2119, 5, 5, 10, 0, 0, tzinfo=UTC),
```
Comment (danjacob-anders) [inline payments/tests/conftest.py]: Is there a reason our reservation is going to be in the 22nd century? (maybe clearer to just get current date and add timedelta to add a few days, but I guess setting a far future date is more efficient).

Comment (jopesy) [reply payments/tests/conftest.py]: No specific reason, I just copied this from another unit test. I can change this, though.

File: payments/tests/test_notifications.py [hunk 3]
```diff
@@ -95,6 +107,23 @@ def test_reservation_created_notification(order_with_products):
...
+@override_settings(RESPA_MAILS_ENABLED=True)
+def test_paid_reservation_approved_notification(requested_reservation_with_order):
+    res = requested_reservation_with_order
+    user = res.user
+    user.preferred_language = 'fi'
+    user.save()
+
+    res.set_state(Reservation.WAITING_FOR_PAYMENT, user)
+
+    assert len(mail.outbox) == 1
```
Comment (danjacob-anders) [inline payments/tests/test_notifications.py]: There's also a `mailoutbox` fixture with pytest django: not sure if easier to use than importing `mail.outbox` separately.

Comment (jopesy) [reply payments/tests/test_notifications.py]: Nice, didn't know about that. Let me try it.

File: resources/api/reservation.py (hunk 1 above)
Comment (jopesy) [reply resources/api/reservation.py]: I was quite sure it would create a circular import but seems not. I'll change this.

Comment (jopesy) [reply resources/api/reservation.py]: Fixed

File: payments/tests/test_notifications.py (hunk 3 above)
Comment (jopesy) [reply payments/tests/test_notifications.py]: Changed. That's quite handy.

File: payments/tests/conftest.py (hunk 2 above)
Comment (jopesy) [reply payments/tests/conftest.py]: Changed the times

**Review by danjacob-anders — APPROVED** [review]
LGTM!

**Pull request merged**